File: MichiMark/Domain/Model/PaymentDomain.swift
import Foundation

struct PaymentDomain: Equatable, Sendable {
    let id: PaymentID

    /// 表示順
    var paymentSeq: Int

    /// 単位: 1円
    var paymentAmount: Int

    /// 支払メンバー（必須）
    var paymentMember: MemberDomain

    /// 割り勘メンバー（空配列可）
    var splitMembers: [MemberDomain]

    /// 入力必須（Stateでは空欄許可）
    var paymentMemo: String?

    /// 論理削除
    var isDeleted: Bool

    /// 登録日（初回のみ設定）
    let createdAt: Date

    /// 更新日（保存時更新）
    var updatedAt: Date

    init(
        id: PaymentID,
        paymentSeq: Int,
        paymentAmount: Int,
        paymentMember: MemberDomain,
        splitMembers: [MemberDomain] = [],
        paymentMemo: String? = nil,
        isDeleted: Bool = false,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.paymentSeq = paymentSeq
        self.paymentAmount = paymentAmount
        self.paymentMember = paymentMember
        self.splitMembers = splitMembers
        self.paymentMemo = paymentMemo
        self.isDeleted = isDeleted
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

File: flutter/lib/domain/transaction/payment/payment_domain.dart
import 'package:equatable/equatable.dart';
import '../../master/member/member_domain.dart';
import 'payment_type.dart';

class PaymentDomain extends Equatable {
  final String id;

  /// 表示順
  final int paymentSeq;

  /// 支払金額（単位: 1円）
  final int paymentAmount;

  /// 支払メンバー（必須）
  final MemberDomain paymentMember;

  /// 割り勘メンバー
  final List<MemberDomain> splitMembers;

  /// メモ（任意）
  final String? paymentMemo;

  /// 論理削除フラグ
  final bool isDeleted;

  /// 登録日（初回のみ設定）
  final DateTime createdAt;

  /// 更新日（保存時更新）
  final DateTime updatedAt;

  /// 紐づく MarkLink の ID。null = PaymentInfo タブから直接登録
  final String? markLinkID;

  /// 支払種別。デフォルト: expense（支出）
  final PaymentType paymentType;

  const PaymentDomain({
    required this.id,
    required this.paymentSeq,
    required this.paymentAmount,
    required this.paymentMember,
    this.splitMembers = const [],
    this.paymentMemo,
    this.isDeleted = false,
    required this.createdAt,
    required this.updatedAt,
    this.markLinkID,
    this.paymentType = PaymentType.expense,
  });

  PaymentDomain copyWith({
    String? id,
    int? paymentSeq,
    int? paymentAmount,
    MemberDomain? paymentMember,
    List<MemberDomain>? splitMembers,
    String? paymentMemo,
    bool? isDeleted,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? markLinkID,
    bool clearMarkLinkID = false,
    PaymentType? paymentType,
  }) {
    return PaymentDomain(
      id: id ?? this.id,
      paymentSeq: paymentSeq ?? this.paymentSeq,
      paymentAmount: paymentAmount ?? this.paymentAmount,
      paymentMember: paymentMember ?? this.paymentMember,
      splitMembers: splitMembers ?? this.splitMembers,
      paymentMemo: paymentMemo ?? this.paymentMemo,
      isDeleted: isDeleted ?? this.isDeleted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      markLinkID: clearMarkLinkID ? null : (markLinkID ?? this.markLinkID),
      paymentType: paymentType ?? this.paymentType,
    );
  }

  @override
  List<Object?> get props => [
        id,
        paymentSeq,
        paymentAmount,
        paymentMember,
        splitMembers,
        paymentMemo,
        isDeleted,
        createdAt,
        updatedAt,
        markLinkID,
        paymentType,
      ];
}
